File: src/mtrtk/core/geo.py
"""WGS84 geodesy: LLH <-> ECEF, UTM (forward), ENU offsets and DMS formatting. Pure functions."""

from __future__ import annotations

import math
from dataclasses import dataclass

WGS84_A = 6378137.0
WGS84_F = 1 / 298.257223563
WGS84_B = WGS84_A * (1 - WGS84_F)
WGS84_E2 = WGS84_F * (2 - WGS84_F)  # first eccentricity squared
_EP2 = WGS84_E2 / (1 - WGS84_E2)  # second eccentricity squared
UTM_K0 = 0.9996
UTM_FALSE_EASTING = 500_000.0
UTM_FALSE_NORTHING_SOUTH = 10_000_000.0


def llh_to_ecef(lat_deg: float, lon_deg: float, h_m: float) -> tuple[float, float, float]:
    lat, lon = math.radians(lat_deg), math.radians(lon_deg)
    sin_lat, cos_lat = math.sin(lat), math.cos(lat)
    n = WGS84_A / math.sqrt(1 - WGS84_E2 * sin_lat * sin_lat)
    x = (n + h_m) * cos_lat * math.cos(lon)
    y = (n + h_m) * cos_lat * math.sin(lon)
    z = (n * (1 - WGS84_E2) + h_m) * sin_lat
    return x, y, z


def ecef_to_llh(x: float, y: float, z: float) -> tuple[float, float, float]:
    lon = math.atan2(y, x)
    p = math.hypot(x, y)
    if p < 1e-9:  # on the polar axis
        lat = math.copysign(math.pi / 2, z)
        return math.degrees(lat), math.degrees(lon), abs(z) - WGS84_B
    lat = math.atan2(z, p * (1 - WGS84_E2))
    for _ in range(20):
        sin_lat = math.sin(lat)
        n = WGS84_A / math.sqrt(1 - WGS84_E2 * sin_lat * sin_lat)
        h = p / math.cos(lat) - n
        new_lat = math.atan2(z, p * (1 - WGS84_E2 * n / (n + h)))
        if abs(new_lat - lat) < 1e-14:
            lat = new_lat
            break
        lat = new_lat
    sin_lat = math.sin(lat)
    n = WGS84_A / math.sqrt(1 - WGS84_E2 * sin_lat * sin_lat)
    h = p / math.cos(lat) - n
    return math.degrees(lat), math.degrees(lon), h


@dataclass(frozen=True)
class Utm:
    zone: int
    hemisphere: str  # "N" | "S"
    easting: float
    northing: float

    @property
    def label(self) -> str:
        return f"{self.zone}{self.hemisphere}"


def utm_zone(lat_deg: float, lon_deg: float) -> int:
    zone = int((lon_deg + 180) // 6) + 1
    if 56 <= lat_deg < 64 and 3 <= lon_deg < 12:
        zone = 32  # south-west Norway
    if lat_deg >= 72 and 0 <= lon_deg < 42:  # Svalbard
        if lon_deg < 9:
            zone = 31
        elif lon_deg < 21:
            zone = 33
        elif lon_deg < 33:
            zone = 35
        else:
            zone = 37
    return min(max(zone, 1), 60)


def llh_to_utm(lat_deg: float, lon_deg: float) -> Utm:
    """Transverse Mercator series (Snyder 1987), accurate to ~1 mm inside the zone."""
    zone = utm_zone(lat_deg, lon_deg)
    hemisphere = "N" if lat_deg >= 0 else "S"
    lat = math.radians(lat_deg)
    lon0 = math.radians((zone - 1) * 6 - 180 + 3)
    d_lon = math.radians(lon_deg) - lon0
    e2, ep2 = WGS84_E2, _EP2
    sin_lat, cos_lat, tan_lat = math.sin(lat), math.cos(lat), math.tan(lat)
    n = WGS84_A / math.sqrt(1 - e2 * sin_lat * sin_lat)
    t = tan_lat * tan_lat
    c = ep2 * cos_lat * cos_lat
    a = d_lon * cos_lat
    m = WGS84_A * (
        (1 - e2 / 4 - 3 * e2**2 / 64 - 5 * e2**3 / 256) * lat
        - (3 * e2 / 8 + 3 * e2**2 / 32 + 45 * e2**3 / 1024) * math.sin(2 * lat)
        + (15 * e2**2 / 256 + 45 * e2**3 / 1024) * math.sin(4 * lat)
        - (35 * e2**3 / 3072) * math.sin(6 * lat)
    )
    easting = (
        UTM_K0
        * n
        * (a + (1 - t + c) * a**3 / 6 + (5 - 18 * t + t * t + 72 * c - 58 * ep2) * a**5 / 120)
    )
    northing = UTM_K0 * (
        m
        + n
        * tan_lat
        * (
            a * a / 2
            + (5 - t + 9 * c + 4 * c * c) * a**4 / 24
            + (61 - 58 * t + t * t + 600 * c - 330 * ep2) * a**6 / 720
        )
    )
    easting += UTM_FALSE_EASTING
    if hemisphere == "S":
        northing += UTM_FALSE_NORTHING_SOUTH
    return Utm(zone, hemisphere, easting, northing)


def ecef_to_enu(
    ref_lat_deg: float, ref_lon_deg: float, ref_h_m: float, x: float, y: float, z: float
) -> tuple[float, float, float]:
    x0, y0, z0 = llh_to_ecef(ref_lat_deg, ref_lon_deg, ref_h_m)
    dx, dy, dz = x - x0, y - y0, z - z0
    lat, lon = math.radians(ref_lat_deg), math.radians(ref_lon_deg)
    sin_lat, cos_lat, sin_lon, cos_lon = math.sin(lat), math.cos(lat), math.sin(lon), math.cos(lon)
    e = -sin_lon * dx + cos_lon * dy
    n = -sin_lat * cos_lon * dx - sin_lat * sin_lon * dy + cos_lat * dz
    u = cos_lat * cos_lon * dx + cos_lat * sin_lon * dy + sin_lat * dz
    return e, n, u


def format_dms(value_deg: float, is_lat: bool, decimals: int = 4) -> str:
    hemi = ("N" if value_deg >= 0 else "S") if is_lat else ("E" if value_deg >= 0 else "W")
    total_seconds = round(abs(value_deg) * 3600, decimals)
    degrees = int(total_seconds // 3600)
    minutes = int((total_seconds - degrees * 3600) // 60)
    seconds = total_seconds - degrees * 3600 - minutes * 60
    width = 3 + decimals if decimals else 2
    return f"{degrees}°{minutes:02d}'{seconds:0{width}.{decimals}f}\"{hemi}"

File: web/src/lib/geo.ts
/**
 * WGS84 geodesy for the browser: LLH <-> ECEF and forward UTM (Snyder 1987 transverse Mercator
 * series). A line-for-line port of `src/mtrtk/core/geo.py`; `geo.test.ts` checks it against
 * `__fixtures__/geo.json`, which that Python module writes (`web/scripts/gen_geo_fixtures.py`).
 * Keep the two in step: a change here without a regenerated fixture fails the tests, and a
 * change to the Python side needs the fixture regenerated and this port re-checked.
 */

export const WGS84_A = 6378137.0;
export const WGS84_F = 1 / 298.257223563;
export const WGS84_B = WGS84_A * (1 - WGS84_F);
/** First eccentricity squared. */
export const WGS84_E2 = WGS84_F * (2 - WGS84_F);
/** Second eccentricity squared. */
const EP2 = WGS84_E2 / (1 - WGS84_E2);
export const UTM_K0 = 0.9996;
export const UTM_FALSE_EASTING = 500_000;
export const UTM_FALSE_NORTHING_SOUTH = 10_000_000;

const rad = (d: number): number => (d * Math.PI) / 180;
const deg = (r: number): number => (r * 180) / Math.PI;

/** Geodetic latitude/longitude (degrees) and ellipsoidal height (m) → ECEF metres. */
export function llhToEcef(latDeg: number, lonDeg: number, hM: number): [number, number, number] {
  const lat = rad(latDeg);
  const lon = rad(lonDeg);
  const sinLat = Math.sin(lat);
  const cosLat = Math.cos(lat);
  const n = WGS84_A / Math.sqrt(1 - WGS84_E2 * sinLat * sinLat);
  return [(n + hM) * cosLat * Math.cos(lon), (n + hM) * cosLat * Math.sin(lon), (n * (1 - WGS84_E2) + hM) * sinLat];
}

/** ECEF metres → [lat°, lon°, ellipsoidal height m]; iterative, converges to 1e-14 rad. */
export function ecefToLlh(x: number, y: number, z: number): [number, number, number] {
  const lon = Math.atan2(y, x);
  const p = Math.hypot(x, y);
  if (p < 1e-9) {
    // on the polar axis
    return [z >= 0 ? 90 : -90, deg(lon), Math.abs(z) - WGS84_B];
  }
  let lat = Math.atan2(z, p * (1 - WGS84_E2));
  for (let i = 0; i < 20; i++) {
    const sinLat = Math.sin(lat);
    const n = WGS84_A / Math.sqrt(1 - WGS84_E2 * sinLat * sinLat);
    const h = p / Math.cos(lat) - n;
    const next = Math.atan2(z, p * (1 - (WGS84_E2 * n) / (n + h)));
    if (Math.abs(next - lat) < 1e-14) {
      lat = next;
      break;
    }
    lat = next;
  }
  const sinLat = Math.sin(lat);
  const n = WGS84_A / Math.sqrt(1 - WGS84_E2 * sinLat * sinLat);
  return [deg(lat), deg(lon), p / Math.cos(lat) - n];
}

/** UTM zone 1..60, with the south-west Norway (32) and Svalbard (31/33/35/37) exceptions. */
export function utmZone(latDeg: number, lonDeg: number): number {
  let zone = Math.floor((lonDeg + 180) / 6) + 1;
  if (latDeg >= 56 && latDeg < 64 && lonDeg >= 3 && lonDeg < 12) zone = 32; // south-west Norway
  if (latDeg >= 72 && lonDeg >= 0 && lonDeg < 42) {
    // Svalbard
    if (lonDeg < 9) zone = 31;
    else if (lonDeg < 21) zone = 33;
    else if (lonDeg < 33) zone = 35;
    else zone = 37;
  }
  return Math.min(Math.max(zone, 1), 60);
}

export interface Utm {
  zone: number;
  hemisphere: "N" | "S";
  /** metres, false easting included */
  easting: number;
  /** metres, false northing included in the south */
  northing: number;
  /** `"46N"` */
  label: string;
}

/** Forward UTM (WGS84), accurate to ~1 mm inside the zone. Raw doubles, like the Python side. */
export function llhToUtm(latDeg: number, lonDeg: number): Utm {
  const zone = utmZone(latDeg, lonDeg);
  const hemisphere: "N" | "S" = latDeg >= 0 ? "N" : "S";
  const lat = rad(latDeg);
  const lon0 = rad((zone - 1) * 6 - 180 + 3);
  const dLon = rad(lonDeg) - lon0;
  const e2 = WGS84_E2;
  const ep2 = EP2;
  const sinLat = Math.sin(lat);
  const cosLat = Math.cos(lat);
  const tanLat = Math.tan(lat);
  const n = WGS84_A / Math.sqrt(1 - e2 * sinLat * sinLat);
  const t = tanLat * tanLat;
  const c = ep2 * cosLat * cosLat;
  const a = dLon * cosLat;
  const m =
    WGS84_A *
    ((1 - e2 / 4 - (3 * e2 ** 2) / 64 - (5 * e2 ** 3) / 256) * lat -
      ((3 * e2) / 8 + (3 * e2 ** 2) / 32 + (45 * e2 ** 3) / 1024) * Math.sin(2 * lat) +
      ((15 * e2 ** 2) / 256 + (45 * e2 ** 3) / 1024) * Math.sin(4 * lat) -
      ((35 * e2 ** 3) / 3072) * Math.sin(6 * lat));
  let easting = UTM_K0 * n * (a + ((1 - t + c) * a ** 3) / 6 + ((5 - 18 * t + t * t + 72 * c - 58 * ep2) * a ** 5) / 120);
  let northing =
    UTM_K0 *
    (m + n * tanLat * ((a * a) / 2 + ((5 - t + 9 * c + 4 * c * c) * a ** 4) / 24 + ((61 - 58 * t + t * t + 600 * c - 330 * ep2) * a ** 6) / 720));
  easting += UTM_FALSE_EASTING;
  if (hemisphere === "S") northing += UTM_FALSE_NORTHING_SOUTH;
  return { zone, hemisphere, easting, northing, label: `${zone}${hemisphere}` };
}
